qiskit/algorithms/minimum_eigen_solvers/qaoa/qaoa_circuit.py

```diff
@@ -0,0 +1,209 @@
+# that they have been altered from the originals.
+
+"""Global X phases and parameterized problem hamiltonian."""
+
+from typing import Optional, Union, cast, Set
+
+import numpy as np
+
+from qiskit import QuantumCircuit
+from qiskit.circuit import Parameter, ParameterVector
+from qiskit.circuit.library import BlueprintCircuit
+from qiskit.opflow import OperatorBase, I, X, CircuitStateFn, H, CircuitOp, EvolutionFactory
+
+
+class QAOACircuit(BlueprintCircuit):
+    """Global X phases and parameterized problem hamiltonian."""
+
+    def __init__(self,
+                 cost_operator: OperatorBase,
+                 param_p: int,
+                 initial_state: Optional[QuantumCircuit] = None,
+                 mixer_operator: Optional[Union[QuantumCircuit, OperatorBase]] = None,
+                 name: str = "qaoa"):
+        """
+        Constructor, following the QAOA paper https://arxiv.org/abs/1411.4028
+
+        Args:
+            cost_operator: The operator representing the cost of the optimization problem,
+                denoted as U(B, gamma) in the original paper.
+            param_p: The integer parameter p, which determines the depth of the circuit,
+                as specified in the original paper.
+            initial_state: An optional initial state to use.
+            mixer_operator: An optional custom mixer to use instead of the global X-rotations,
+                denoted as U(B, beta) in the original paper. Can be an operator or
+                an optionally parameterized quantum circuit.
+            name: A name of the circuit, default 'qaoa'
+        Raises:
+            TypeError: invalid input
+        """
+        super().__init__(name=name)
+        self._cost_operator = cost_operator
+        self._param_p = param_p
+        self._initial_state = initial_state
+        self._mixer_operator = mixer_operator
+
+        # set this circuit as a not-built circuit
+        self._num_qubits = 0
+        self._num_parameters = 0
+        self._bounds = None
+        self._mixer = None
+        self._data = None
+
+    def _check_configuration(self, raise_on_failure: bool = True) -> bool:
+        valid = True
+
+        if self._cost_operator is None:
+            valid = False
+            if raise_on_failure:
+                raise AttributeError("The operator representing the cost of "
+                                     "the optimization problem is not set")
+        if self._param_p is None or self._param_p < 1:
+            valid = False
+            if raise_on_failure:
+                raise AttributeError("The integer parameter p, which determines the depth "
+                                     "of the circuit, either not set or set to non-positive value")
+
+        return valid
+
+    def _build(self) -> None:
+        """Build the circuit."""
+        if self._data:
+            return
+
+        self._check_configuration()
+        self._data = []
```

Comment:
Since this logic is implemented as build in BlueprintCircuit is there a reason not to simply call super()._build() to do this. Having looked at what NLocal does I see it replicates the logic in its overridden build.

Reply:
Plain call to `super()._build()` does not work. If `self._data is not None` is true then the super method would just return and `qaoa.build()` would continue while it should be stopped. So, another if statement is required anyway, then why call super?

Reply:
The rationale, in general, is to let the parent enact whatever it needs to do at build time without necessarily needing knowledge of what its doing, then once that is done do whatever you need to do as derived class. This way if the parent adds any extra stuff that needs doing it at its level it gets done by calling super.

Reply:
Hm are you sure it doesn't work? Most other classes do call the super method. If you look at the code then `super()._build()` does exactly what Steve marked above. I think you should replace all those lines with just `super()._build()`. (Of course calling super and then doing the checks again doesn't work 🙂)

Reply:
Alternative is:
```python
    def _build(self) -> None:
        """Build the circuit."""
        super()._build()
        if self._data:
            return
```
This works, without this `if` statement does not as I wrote above.